In [1]:
import prepare
import project_acquire

In [11]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import nltk
import unicodedata
import re
import numpy as np


In [3]:
df=prepare.wrangle_data()

In [8]:
df

repo    language  \
0                            rdpeng/ExData_Plotting1        None   
1                     rdpeng/RepData_PeerAssessment1        None   
2                  DataScienceSpecialization/courses        HTML   
3                              fighting41love/funNLP      Python   
4                                   magento/magento2         PHP   
..                                               ...         ...   
485                                   rprokap/pset-9  JavaScript   
486                                 konzy/mass_clone       Shell   
487                               biter777/countries          Go   
488  sayantann11/all-classification-templetes-for-ML      Python   
489                                  denman2328/Help        None   

                                       readme_contents  \
0    ## Introduction\n\nThis assignment uses data f...   
1    ## Introduction\n\nIt is now possible to colle...   
2    \n### Data Science Specialization\n\nThese are...   
3    <center>\n    <img style="border-radius: 0.312...   
4    \n<p align="center">\n<a href="https://www.cod...   
..                                                 ...   
485  # pset-9\n      CREDITS SEQUENCE              ...   
486  # mass_clone\nThis is a shell script that will...   
487  # countries\r\n\r\nCountries - ISO 3166 (ISO31...   
488  # all-classification-templetes-for-ML\nClassif...   
489  ------------------\nSystem Information\n------...   

                                               stemmed  \
0    introductionthi assign use data fromth uc irvi...   
1    introductionit possibl collect larg amount dat...   
2    data scienc specializationthes cours materi jo...   
3    nlpdatalogocitations487redsvgdatalogohomee4bab...   
4    magento open sourcewelcom magento open sourc p...   
..                                                 ...   
485  pset9credit sequencenewspap headlin montagehea...   
486  massclonethi shell script clone multipl reposi...   
487  countriescountri iso 3166 iso31661 iso3166 dig...   
488  allclassificationtempletesformlclassif machin ...   
489  system informationtim report 8102013 083620 ma...   

                                            lemmatized  
0    introductionthis assignment us data fromthe uc...  
1    introductionit possible collect large amount d...  
2    data science specializationthese course materi...  
3    nlpdatalogocitations487redsvgdatalogohomee4bab...  
4    magento open sourcewelcome magento open source...  
..                                                 ...  
485  pset9credits sequencenewspaper headline montag...  
486  massclonethis shell script clone multiple repo...  
487  countriescountries iso 3166 iso31661 iso3166 d...  
488  allclassificationtempletesformlclassification ...  
489  system informationtime report 8102013 083620 m...  

[490 rows x 5 columns]

# Questions to answer:

## What are the most common words in READMEs?


## Does the length of the README vary by programming language?

In [9]:
df.columns

Index(['repo', 'language', 'readme_contents', 'stemmed', 'lemmatized'], dtype='object')

In [15]:
# breaking down the DataFrame by language, its n number of occurances, and percentage of the data

language = pd.concat([df.language.value_counts(),
                    df.language.value_counts(normalize=True)], axis=1)
language.columns = ['n', 'percent']
language


n   percent
Python            71  0.194521
Java              66  0.180822
JavaScript        55  0.150685
Assembly          39  0.106849
HTML              22  0.060274
C++               21  0.057534
PHP               16  0.043836
C#                14  0.038356
C                 12  0.032877
Shell              9  0.024658
Jupyter Notebook   6  0.016438
CSS                4  0.010959
PowerShell         4  0.010959
Go                 4  0.010959
TSQL               3  0.008219
Dockerfile         2  0.005479
R                  2  0.005479
Swift              2  0.005479
Vue                2  0.005479
Ruby               1  0.002740
Objective-C++      1  0.002740
Julia              1  0.002740
Emacs Lisp         1  0.002740
Objective-C        1  0.002740
Kotlin             1  0.002740
Jinja              1  0.002740
SCSS               1  0.002740
Brainfuck          1  0.002740
Rich Text Format   1  0.002740
VHDL               1  0.002740

In [19]:
# separating the README corpus by language classification. 

all_words = ' '.join(df.lemmatized)
python_words =' '.join(df[df.language == 'Python'].lemmatized)
java_words =' '.join(df[df.language == 'Java'].lemmatized)
java_script_words =' '.join(df[df.language == 'JavaScript'].lemmatized)


In [26]:
# taking each of the word lists that are separated by language (except for all_freq variable which looks
# at all the README files combined as a whole) and finding the most frequently used words

all_freq = pd.Series(all_words.split()).value_counts()
python_freq = pd.Series(python_words.split()).value_counts()
java_freq = pd.Series(java_words.split()).value_counts()

all_freq.head(20)


'              74348
0              37602
gem5           26604
1              21658
instruction    20649
kernel         20166
arch           17361
see            17189
also           16996
run            15784
aarch64        15088
example        13937
one            13646
arm            13349
use            13314
qemu           13173
movz           12423
event          12238
build          11869
get            11392
dtype: int64

### the top words for all README files are listed above

# second question

## Does the length of the README vary by programming language?

In [43]:

# adding a new coloumn to the dataframe "readme_length" that displays the word count of the README file content or the 
# corresponding row

df['readme_length']=df.lemmatized.str.split().str.len()

In [44]:
# making sure the output looks correct for the new column
df.readme_length

0        320
1        646
2         36
3       1072
4        334
       ...  
485    12801
486      106
487      536
488     1880
489     4758
Name: readme_length, Length: 490, dtype: int64

In [45]:
# breaking down the dataframe by language again and this time computing the average README content word count for each

df.groupby('language')['readme_length'].mean()

language
Assembly             4216.769231
Brainfuck             244.000000
C                     281.833333
C#                    842.785714
C++                   132.761905
CSS                   167.250000
Dockerfile            305.000000
Emacs Lisp            473.000000
Go                    676.750000
HTML                  109.136364
Java                  466.939394
JavaScript           8081.327273
Jinja                 270.000000
Julia                 236.000000
Jupyter Notebook     1071.833333
Kotlin                 47.000000
Objective-C           401.000000
Objective-C++         483.000000
PHP                   238.750000
PowerShell            735.750000
Python              40471.309859
R                     998.000000
Rich Text Format      101.000000
Ruby                   89.000000
SCSS                   92.000000
Shell                 218.333333
Swift                 377.500000
TSQL                    0.000000
VHDL                  288.000000
Vue                    27.500000
N

# looking at the above output there are definitely some outliers when it comes to README  word counts. Python has a staggering 40,471 average, javascript is just over 8,000, assembly is 4,266.  The rest really coalesce and have similar aggregation amounts in the hundreds